File: FindMinRotatedSortedArray.py
"""
Find Minimum in Rotated Sorted Array

Brute force approach -> Linear search (TC - O(n) OR sort the array and return 0th index element (TC - O(n log n))

Optimized BS approach:
TC - O(log n)
SC - O(1)
"""

nums = [4, 5, 6, 7, 0, 1, 2]

# nums = [2, 1]

class Solution:
    def findMin(self, nums):
        # Base case
        if nums is None or len(nums) == 0: return -1

        n = len(nums)

        low = 0
        high = n - 1

        while low < high:
            # check if array is already sorted
            if nums[low] < nums[high]: return nums[low]

            mid = low + (high - low) // 2

            # check boundary condition and check if mid is the min
            if (mid == 0 or nums[mid] < nums[mid - 1]) and (nums[mid] < nums[mid + 1]):
                return nums[mid]
            elif nums[mid] < nums[high]:
                # search in left
                high = mid - 1
            else:
                low = mid + 1
        return nums[low]


obj = Solution()
print(obj.findMin(nums))

File: FindMinRotatedSortedArray.java
// Approach: Two violations of sortedness occur when an element is smaller than its preceding element or greater than its following
// element. By modifying the binary search return condition to check for these violations and reducing the search space based on the
// sorted side of the midpoint, this problem can be solved efficiently. It's similar to finding pivot index in a sorted rotated array.
// Returning the first element as minimum when the array is unrotated is an important edge case to consider.
// Time Complexity : O(logn)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

public class FindMinRotatedSortedArray {

    int findMin(int[] arr) {
        int low = 0, high = arr.length - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (mid < arr.length - 1 && arr[mid] > arr[mid + 1]) {
                return mid + 1;
            } else if (mid > 0 && arr[mid] < arr[mid - 1]) {
                return mid;
            }
            if (arr[mid] >= arr[low]) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return 0;
    }

    public static void main(String[] args) {
        FindMinRotatedSortedArray fmrsa = new FindMinRotatedSortedArray();
        int[] arr = { 4, 5, 6, 7, 0, 1, 2 };
//        int[] arr = { 1, 2, 3, 4, 5 };
        int idx = fmrsa.findMin(arr);
        System.out.println("Minimum element is " + arr[idx] + " and it's found at index " + idx);
    }
}